pyrevolve/pyrevolve.py

```diff
@@ -115,107 +212,354 @@ def apply_reverse(self):
         forward run."""
 
-        with self.profiler.get_timer('reverse', 'reverse'):
-            self.rev_operator.apply(t_start=self.scheduler.capo,
-                                    t_end=self.scheduler.capo+1)
-        while(True):
+        action = None
+        remove_ckp_flag = False
+        while True:
             # ask Revolve what to do next.
-            action = self.scheduler.next()
-            if(action.type == Action.ADVANCE):
-                # advance forward computation
-                with self.profiler.get_timer('reverse', 'advance'):
-                    self.fwd_operator.apply(t_start=self.scheduler.old_capo,
-                                            t_end=self.scheduler.capo)
-            elif(action.type == Action.TAKESHOT):
+            if remove_ckp_flag is True:
+                action = HAction(
+                    action_type=Action.CPDEL,
+                    capo=self.__last_capo_read,
+                    index=[self.__last_stidx_read],
+                    old_capo=self.scheduler.old_capo,
+                    ckp=self.scheduler.cp_pointer,
+                )
+                remove_ckp_flag = False
+            else:
+                action = self.scheduler.next()
+
+            if action.type == Action.REVERSE:
+                # advance adjoint computation by a single step
+                with self.profiler.get_timer("reverse", "reverse"):
+                    self.fwd_operator.apply(
+                        t_start=self.scheduler.capo, t_end=self.scheduler.capo + 1
+                    )
+                    self.rev_operator.apply(
+                        t_start=self.scheduler.capo, t_end=self.scheduler.capo + 1
+                    )
+                if self.scheduler.capo == self.__last_capo_read:
+                    # h-revolve implies that B^i must also remove
+                    # the checkpoint if it was previously restored
+                    remove_ckp_flag = True
+            elif action.type == Action.REVSTART:
+                """Sets the rev_operator to 'nt' only if its not already there.
+                This condition happens when using CRevolve shceduler, but not
+                when using HRevolve.
+                """
+                with self.profiler.get_timer("reverse", "reverse"):
+                    self.rev_operator.apply(
+                        t_start=self.scheduler.capo, t_end=self.scheduler.capo + 1
+                    )
+            elif action.type == Action.TAKESHOT:
                 # take a snapshot: copy from workspace into storage
-                with self.profiler.get_timer('reverse', 'takeshot'):
-                    self.save_checkpoint()
-            elif(action.type == Action.RESTORE):
+                with self.profiler.get_timer("reverse", "takeshot"):
+                    self.save_checkpoint(action.storageIndex())
+
+                    if (self.scheduler.capo == self.__last_capo_read) and (
```

Comment:
I am not enjoying seeing this kind of code here. It looks so specialised to one scheduler.